src/server/project.ts

```diff
@@ -834,6 +846,14 @@ namespace ts.server {
             (this.updatedFileNames || (this.updatedFileNames = createMap<true>())).set(fileName, true);
         }
 
+        /*@internal*/
+        markFileAsDirty(changedFile: string) {
+            this.markAsDirty();
+            if (!this.importSuggestionsCache.isEmpty) {
+                (this.dirtyFilesForSuggestions || (this.dirtyFilesForSuggestions = createMap())).set(changedFile, true);

```

Comment:
why do we need to track these ? Can we not clean cache right here instead ?

Reply:
I believe at this point we don’t have the new source file parsed yet. To determine if the cache needs to be cleared after editing a source file, we need to have both the old source file and the new one together so we can run `this.sourceFileHasChangedOwnImportSuggestions(oldSourceFile, sourceFile)`. Also, that analysis can be skipped for some values of `oldProgram.structureIsReused`, so we really have to wait until we have a new program.

Reply:
So currently we have `finishRecordingFilesWithChangedResolutions` that gets you result of changed unresolved imports.. You could similarly create and get result of files with changed resolution from resolutionCache instead of doing this work again.. If fact you need `boolean` so you can clear import cache..

Reply:
Changed resolutions isn’t quite the same thing we’re detecting here. For example, given two files:

```ts
// a.ts
import { useContext } from 'react';
```

```ts
// b.ts
useContext/**/ // Auto-import from 'react'
```

If the user accepts an auto-import from 'react' in b.ts, that file has changed resolution, but the program didn’t change in a way that could possibly affect the auto-imports available in b.ts. _Most_ changes to a given file don’t affect its _own_ auto-imports. The ones that do are

- Importing something that causes a change to the program structure (typically bringing new files into the program)
- Creating or changing a module augmentation or ambient module declaration
- Adding the first or removing the last import of any node core module in a file, when ATA is enabled

I’m not 100% sure if I’m understanding your suggestion correctly, but as to the concern about doing work again, I found no measurable impact on the time reported by `updateGraphWorker`.

Reply:
What i mean is dirtyFilesForSuggestions array isnt it same or subset of filesWithChangedResolution? So there is no need to keep this array is all i am saying..

Reply:
> isnt it same or subset of filesWithChangedResolution

No, it’s a superset. I tried your suggestion just to make sure, and three cache invalidation tests failed. One example is `importSuggestionsCache_ambient.ts`:

```ts
declare module 'ambient' {
  export const ambient = 0;
}
a/**/
```

After an edit that deletes `export const ambient = 0;`, it still shows up in completions.

I would be happy to consider another way of getting the list of changed files, but it really has to be all files that changed at all.

Reply:
what about all files with `oldProgram.getSourceFIle().version !== newProgram.getSourceFile().version` => These should definitely be all dirtyFiles?

Reply:
Yeah, that would definitely work, I was just trying to avoid iterating over all source files again when there’s already a callback from ScriptInfo to Project every time a file changes.

Reply:
I measured this and iterating through all source files on a large real-world project adds ~10 ms to updateGraphWorker, which accounts for about 10%. That doesn’t seem worth it.